### ISSUE TITLE
Major - Dashboard: Results not displaying

### ISSUE THREAD

**Issue by T-T-Nguyen** [StatCan/WellbeingCheckUAT#290]
User cannot access the dashboard after 4 cases - Error message requesting user to complete 2 or more questionnaires to obtain results.

IOS 15.0.2| EN	| iPhone Xr

- Reinstalled on Oct 25
- Submitted 4 cases, even 1 hour later still can't see the dashboard

Please look into the backend to see what had happened.
Whether or not this can be avoided or resolved, some background information would be appreciated should additional information is required for the Help Desk.

-

**Comment by thierryrama** [comment]
@T-T-Nguyen, do you know the timeframe when the cases were submitted so that I can narrow down my search in the backend?

**Comment by Jiandong-Guo** [comment]
Is the submitted 4 cases are all survey A or 1 survey A + 3 survey B ?

**Comment by Jiandong-Guo** [comment]
2 options to try:
1) if the user know how to kill background running app, then kill Vitali-T app and run it again, and do a survey B, check result.
    (Kill background running app:  double tag phone home button, you will see all the background running apps in small square icon, move Vitali-T out the screen)

2)If user doesn't know how to kill background running app, then reboot the phone, do a survey B, check result

**Comment by zhuelle** [comment]
@T-T-Nguyen would you please update how the user gets the dashboard working? did the tester do nothing and it worked or the tester did another reinstall and it worked? This is all information the dev team required for further investigation.

**Comment by T-T-Nguyen** [comment]
@zhuelle As mentioned, both testers we able to view the dashboards later in the day.  Perhaps this is the slowness in the server as stated yesterday. Is Thierry or someone investigating?

Julie - was able to see it around 3pm the day of*

**Comment by zhuelle** [comment]
@T-T-Nguyen do you mean both testers completed 4 surveys, not able to click the Report button. Later the same day, they logged into the app and were able to click the Report button and see the report?

**Comment by T-T-Nguyen** [comment]
Yes, Julie completed submitted 4 surveys (B) and was unable to view her results on Monday morning. Around 3pm she was able to view the associated dashboards through the results button.

**Comment by zhuelle** [comment]
After verifying, it was possible to isolate Julie’s testing on Oct 25th in the logs and noticed a strange behavior from the WebAPI and requested new logs from IIS hosting:
•	Around 9:58AM she started with a new install and completed Questionnaire A at 10:00AM
•	The mobile app has set the password to the server at 10:01:00
•	Started Questionnaire B at 10:00:59 and completed at 10:01:55
•	Started another Questionnaire B at 10:28:31 and completed at 10:30:46
•	The mobile app requested the graph at 10:30:52 but the WebAPI login to get the JWT token returned a HTTP code 0 and no token were returned.
o	The mobile app could not then get the graphs because there was no tokens
•	Started another Questionnaire B at 11:06:05 and completed at 11:06AM
•	The mobile app tried to retrieve the graphs with an empty token at 11:08:09 and denied
•	Started another Questionnaire B at 11:08:09 but did not complete (this may not be accurate as I am tracking by IP address so may have changed temporarily)
•	Started another Questionnaire B at 11:54:54 but did not complete (this may not be accurate as I am tracking by IP address so may have changed temporarily)
•	Started another Questionnaire B at 17:00:23 and completed at 17:00:39
•	The mobile app requested the graph at 17:00:41 and the WebAPI always returned 200 so it was successful.

The cause and possible fix: 

•	An error on the server side prevented the dashboard from being downloaded
•	The issue will resolve by itself when the respondent clicks the Report button in different time when sever doesn’t have this issue
•	A possible fix on the mobile app side is to add a check for the sever issue and retry a couple times more to get the graph

**Comment by T-T-Nguyen** [comment]
@zhuelle Please add On Hold label
To be re-evaluated

**Comment by thierryrama** [comment]
The server-side errors (HTTP code 0 and 500) associated with this issue is now being seen on the servers more often now (many times a day) that we have more respondents. The symptoms would be missing graphs or failure to reset password.

Analyzing the logs indicates that it happens when a particular server (we have 37 servers) was idled for more than 20 minutes (the exact time out is in the server configuration) and received a request. The request takes sometime to respond and the mobile app just disconnects (seen on the server as a connection reset and thus an operation cancel)

When an application (WebAPI or EQ)  on the server is idled (not receiving requests) for more than 20 minutes, it is removed from memory. Upon the first request after idling, it is brought back to memory but that operation takes time which may trigger a disconnect from the mobile app side.

We are asking IIS hosting if there is a way to keep the application in memory longer (or permanently).

**Comment by T-T-Nguyen** [comment]
@thierryrama Please keep up posted- Thank you

**Comment by thierryrama** [comment]
IIS hosting is looking into what is possible. It is technically feasible but they have to evaluate the impact on their hosting environments.

**Comment by thierryrama** [comment]
I gathered some stats from the logs for **Nov 15th** to put the issue into perspective:
Total requests to the server: **8289**
Error HTTP 500 or 0: **133**

The error rate for this issue is below 2%. It seems to be the same rate for the last 5 days. So we will get those errors everyday. This is why we did not see much of the errors during testing because we did not have sufficient number of requests.

The impact on users is that they may fail 
- to get one graph or all graphs that one time. Next time they may get lucky
- to reset the password

@T-T-Nguyen do you want us to pursue trying to fix the issue? It will require a release on the server side (not the app store). We will do a QA release before attempting Prod. We will need to seek approval from ICOS (timing with LFS) as well for the release to Prod.

**Comment by T-T-Nguyen** [comment]
@thierryrama Will get back to you - thank you for the update.

**Comment by T-T-Nguyen** [comment]
@thierryrama Please proceed with the change after this month's LFS cycle.

**Comment by thierryrama** [comment]
For Nov 17th,
Total requests: 11729
Error HTTP 500 or 0: 130

Number of requests are going up (more people) but the error stayed stable. 
The release https://jirab.statcan.ca/browse/ICOS-2457 will hopefully bring it close to 0.

**Comment by thierryrama** [comment]
Fix was deployed to Prod on Nov 26th at 7AM.
For **Nov 26th**
Total requests: 14682
Error HTTP 500 or 0: 59

For **Nov 27th**
Total requests: 14830
Error HTTP 500 or 0: 12

This is a big improvement while not bringing the number to 0. The server is way more responsive at any time of the day compared to before.

Why the error did not go to 0?
To be conservative, we allowed the application on the server to recycle every 6 hours. This is to mitigate any unforeseen memory leak. The HTTP 500 or 0 error now only happens around the time (7AM, 1PM, 7PM, 1AM) the app on the server recycles.

Given the low number, we will just keep it like that without further change.
